internal/vcs/git/commits.go

```
@@ -387,6 +390,7 @@ func CommitCount(ctx context.Context, repo api.RepoName, opt CommitsOptions) (ui
 }
 
 // FirstEverCommit returns the first commit ever made to the repository.
+// TODO: sub-repo filtering
```

I'm not sure what we want to do here. Here are the options I've thought of:
1) return the sha for the actual first commit, even if the user doesn't have access to some files modified in that commit. If they try to fetch any information about the commit, they'll get a `RevisionNotFound` error due to the sub-repo permissions filtering on `GetCommit` anyway, so maybe it doesn't matter?
2) We call the `git rev-list --max-count=1 --max-parents=0 HEAD` command and pass the resulting commit to `GetCommit` as is the current behavior. If `GetCommit` returns a `RevisionNotFound` (i.e. we don't have access), we check the next commit after that and continue until we find first commit the user has access to. This could be bad if there are a ton of commits or if the git history is really complicated. I'm not entirely sure what it'd look like.
3) List all commits (w/ sub-repo permissions filtering, so all commits listed are ones the user has access to) and take the earliest commit. I also don't know if this will give us what we want.

Ultimately, it would be helpful to know how this is used to determine what should be returned here. Very open to any suggestions or insight.

---

Looks like this is only used by query insights. cc @coury-clark

---

Thanks @keegancsmith for the ping.

Just strictly from a Code Insights perspective, this is used internally and will likely always be called as an internal actor - for which we would expect to receive *all* the results.

Having said that, a function named `FirstEverCommit` I would expect to either return exactly what it says, or otherwise return an error if for some reason that isn't possible. I think that filtering on top of this will lead to confusion / problems in ways we could never anticipate, and I don't suggest that.

---

Sounds good to me. I think we can just return an error if the user doesn't have access. The internal actor should always have access, so the error shouldn't ever pop up, but if someone decides to use this function elsewhere we'd at least have a layer of protection.